File: app/src/main/java/org/weishe/kotlindemo/http/BaseObserver.java
package org.weishe.kotlindemo.http;

import com.google.gson.JsonParseException;

import org.json.JSONException;

import java.io.InterruptedIOException;
import java.net.ConnectException;
import java.net.UnknownHostException;
import java.text.ParseException;

import io.reactivex.observers.DisposableObserver;
import retrofit2.HttpException;

public abstract class BaseObserver<T> extends DisposableObserver<BaseResponse<T>> {


    private boolean isShowDialog;

    @Override
    protected void onStart() {

    }

    @Override
    public void onNext(BaseResponse<T> response) {

    }

    @Override
    public void onError(Throwable e) {

        BaseException be;

        if (e != null) {
            //自定义异常
            if (e instanceof BaseException) {
                be = (BaseException) e;

            } else {
                if (e instanceof HttpException) {
                    //HTTP错误
                    be = new BaseException(BaseException.BAD_NETWORK_MSG, e);
                } else if (e instanceof ConnectException || e instanceof UnknownHostException) {
                    //连接错误
                    be = new BaseException(BaseException.CONNECT_ERROR_MSG, e);
                } else if (e instanceof InterruptedIOException) {
                    //连接超时
                    be = new BaseException(BaseException.CONNECT_TIMEOUT_MSG, e);
                } else if (e instanceof JsonParseException || e instanceof JSONException || e instanceof ParseException) {
                    //解析错误
                    be = new BaseException(BaseException.PARSE_ERROR_MSG, e);
                } else {
                    be = new BaseException(BaseException.OTHER_MSG, e);
                }
            }
        } else {
            be = new BaseException(BaseException.OTHER_MSG);
        }
        onError(be.getErrorMsg());
    }

    @Override
    public void onComplete() {

    }


    public abstract void onSuccess(T o);

    public abstract void onError(String msg);

}

File: app/src/main/java/org/weishe/kotlindemo/http/BaseObserver.kt
package org.weishe.kotlindemo.http

import com.google.gson.JsonParseException
import io.reactivex.observers.DisposableObserver
import org.json.JSONException
import retrofit2.HttpException
import java.io.InterruptedIOException
import java.net.ConnectException
import java.net.UnknownHostException
import java.text.ParseException

abstract class BaseObserver<T> : DisposableObserver<BaseResponse<T>?>() {
    private val isShowDialog = false
    override fun onStart() {}
    override fun onNext(response: BaseResponse<T>) {}
    override fun onError(e: Throwable) {
        val be: BaseException = if (e is BaseException) {
            e
        } else {
            if (e is HttpException) { //HTTP错误
                BaseException(BaseException.BAD_NETWORK_MSG, e)
            } else if (e is ConnectException || e is UnknownHostException) { //连接错误
                BaseException(BaseException.CONNECT_ERROR_MSG, e)
            } else if (e is InterruptedIOException) { //连接超时
                BaseException(BaseException.CONNECT_TIMEOUT_MSG, e)
            } else if (e is JsonParseException || e is JSONException || e is ParseException) { //解析错误
                BaseException(BaseException.PARSE_ERROR_MSG, e)
            } else {
                BaseException(BaseException.OTHER_MSG, e)
            }
        }
        onError(be.errorMsg)
    }

    override fun onComplete() {}
    abstract fun onSuccess(o: T)
    abstract fun onError(msg: String?)
}